ocaml/xapi/xapi_database_backup.ml

```diff
@@ -0,0 +1,184 @@
+            Xapi_slave_db.slave_db := Db_cache_types.remove_row tblname objref !Xapi_slave_db.slave_db
+          else
+            Printf.printf "Not valid ref"
+       ) dtbl.table dtbl.key
+    ) deletes
+
+let make_change __context mtime tblname objref fldname newval =
+  Xapi_slave_db.slave_db :=
+    Database.update (
+      (fun _ -> newval)
+      |> Row.update mtime fldname newval
+      |> Table.update mtime objref Row.empty
+      |> TableSet.update mtime tblname Table.empty) !Xapi_slave_db.slave_db
+
+let unpack_row __context (table:table) tblname row =
+  List.iter (fun (f:field)
+              -> Mutex.execute Xapi_slave_db.slave_db_mutex
+                  (fun () -> make_change __context f.stat.modified tblname row.objref f.fldname f.value)
+            ) row.value
+
+let unpack_table __context (t:table) =
+  List.iter (unpack_row __context t t.tblname) (if t.value = [] then [empty_row] else t.value)
+
+let apply_updates __context updates =
+  List.iter (unpack_table __context) updates
+
+let get_table_list db =
+  Db_cache_types.TableSet.fold_over_recent (-2L) (fun table stat value acc -> table::acc) (Db_cache_types.Database.tableset db) []
+
+let counter db table =
+  let t = Db_cache_types.TableSet.find table (Db_cache_types.Database.tableset db) in
+  {tblname=table; count=0; del=(Db_cache_types.Table.get_deleted_len t)}
+
+let object_count db =
+  let tables = get_table_list db in
+  List.map (counter db) tables
+
+let count_eq_check c1 c2 =
+  c1.tblname = c2.tblname && c1.count = c2.count && c1.del = c2.del
+
+let apply_changes (delta:delta) __context =
+  if (Db_cache_types.Manifest.generation (Db_cache_types.Database.manifest !Xapi_slave_db.slave_db)) < delta.fresh_token then
+    begin
+      Xapi_slave_db.slave_db := Database.set_generation delta.fresh_token !Xapi_slave_db.slave_db;
+      apply_updates __context delta.tables;
+      apply_deletes __context delta.deletes;
+    end;
+  (* Need to check that we haven't missed anything *)
+  if not (List.for_all (fun x -> x)
+      (List.map2 count_eq_check (object_count !Xapi_slave_db.slave_db) (object_count (Db_ref.get_database (Context.database_of __context))))) then
+    Mutex.execute Xapi_slave_db.slave_db_mutex (fun () -> Xapi_slave_db.slave_db := Db_cache_types.Database.set_generation 0L !Xapi_slave_db.slave_db)
+
+let info_extract (tlist:table list) =
+  let tbls = List.map (fun (i:table) -> i.tblname) tlist in
+  let rows = List.concat (List.map (fun tbl -> List.map (fun row -> row.objref) tbl.value) tlist) in
+  List.concat (List.map ((fun objlist tblname -> List.combine [tblname] objlist) rows) tbls)
+
+let handle_notifications __context (d:delta) =
+  List.iter ((fun __context (d:del_tables) -> send_notification __context "del" d.table d.key) __context) d.deletes;
+  let tbls, objs = List.split (info_extract d.tables) in
+  List.iter2 (send_notification __context "mod") tbls objs
```

Comment:
Could this and info_extract be written with a `fold` instead? It looks like we are constructing a lot of intermediate values, only to iterate over them.